File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/1360-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    public int maxLength(List<String> arr) {
        int  n = arr.size();
        String temp = "";
        return func(arr , 0 , n , temp);
    }
    int func(List<String> arr , int idx , int n , String temp){
        if( idx >= n){
            return temp.length();
        }
        int include = 0 ,exclude = 0;
        if(hasDuplicates(arr.get(idx) , temp)){
            exclude = func(arr , idx+1, n , temp);
        }else{
            exclude = func(arr, idx+1, n , temp);
            include = func(arr , idx+1 , n , temp+arr.get(idx));
        }
        return Math.max(include , exclude);
    }
    boolean hasDuplicates(String s1 , String s2){
        boolean v[] = new boolean[26];
        for(int i =0 ; i<s1.length() ; i++){
            if(v[s1.charAt(i) -'a'] == true){
                return true;
            }
            v[s1.charAt(i) -'a'] = true;
        }

        for(int i =0 ; i<s2.length() ; i++){
            if(v[s2.charAt(i) -'a'] == true){
                return true;
            }
            v[s2.charAt(i) -'a'] = true;
        }
        return false ;
    }
}

File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/1360-maximum-length-of-a-concatenated-string-with-unique-characters.go

var memo map[string]int

func maxLength(arr []string) int {
	memo = make(map[string]int)
	return fun(arr, 0, "")
}

func fun(arr []string, idx int, temp string) int {
	if idx >= len(arr) {
		return len(temp)
	}
	key := fmt.Sprintf("%d,%s", idx, temp)
	if val, found := memo[key]; found {
		return val
	}

	exclude := fun(arr, idx+1, temp)
	include := 0
	if !hasDuplicates(arr[idx], temp) {
		include = fun(arr, idx+1, temp+arr[idx])
	}
	result := max(exclude, include)
	memo[key] = result
	return result
}

func hasDuplicates(s1, s2 string) bool {
	v := make([]bool, 26)
	for _, char := range s1 {
		if v[char-'a'] {
			return true
		}
		v[char-'a'] = true
	}
	for _, char := range s2 {
		if v[char-'a'] {
			return true
		}
		v[char-'a'] = true
	}
	return false
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
